### ISSUE TITLE
💡 [REQUEST] - Dockerfile from docker image tarball without dependency on docker

### ISSUE THREAD

**Issue by nightlark** [LanikSJ/dfimage#253]
### Start Date

05/07/2024

### Implementation PR

Not yet.

### Summary

I was reading the Docker container spec, and it looks like extracting history information from a Docker image that has been saved as a tar file would primarily involve reading a JSON file and pulling information from a few fields. Using this could enable re-creating the Dockerfile commands used to create a Docker image that has been exported as a tar file, without re-importing it into docker. Not depending on having a docker unix socket available would also enable this functionality to work on systems using alternative container runtimes (e.g. podman) that may not expose a docker compatible unix socket.

### Basic Example

Basically, enable running dfimage on a Docker image that has been saved/exported to a tar file -- without any dependency on having `docker` installed (or communicating with docker). e.g. `dfimage saved_docker_image.tar` should be able to reproduce the sequence of dockerfile commands used to create it.

### Drawbacks

Getting local tag names associated with image IDs referenced wouldn't be possible if docker isn't installed -- however, if docker is present, the existing method of calling out to Docker over the unix socket would still be used; there shouldn't be any impact on existing use cases.

**Comment by redtux** [comment]
Hi @nightlark,

have you tried running podman in privileged mode?

✅  This works fine for me (with podman version 4.3.1 on Ubuntu 23.10 x86_64):

```bash
sudo podman run -v /run/podman/podman.sock:/var/run/docker.sock ghcr.io/laniksj/dfimage docker.io/library/ruby:latest
```

❎ Running as unprivileged user didn't work:

```bash
podman run -v /run/user/$(id -ur)/podman/podman.sock:/var/run/docker.sock ghcr.io/laniksj/dfimage docker.io/library/ruby:latest
```

💡 Here is the error I get in case of interest:

```py
Traceback (most recent call last):
  File "/app/entrypoint.py", line 126, in <module>
    __main__ = MainObj()
               ^^^^^^^^^
  File "/app/entrypoint.py", line 23, in __init__
    self._get_layers_with_images()
  File "/app/entrypoint.py", line 69, in _get_layers_with_images
    self.layers_with_images[last_layer_id] = i["RepoTags"][0]
                                             ~~~~~~~~~~~~~^^^
TypeError: 'NoneType' object is not subscriptable
```

@LanikSJ: Maybe this should be added to the readme? Great project btw. 👏🏼

**Comment by LanikSJ** [comment]
Thank you 👍 so much @redtux I'll add it to the Readme. I'm also thinking of adding a wiki for some of the more common questions / answers. I'll try and get to it some time soon.

**Comment by redtux** [comment]
Wow, this was really fast! 🥇 

Regarding the initial question, this could also be a workaround if running podman as root is not option:

```bash
podman history --no-trunc --format {{.CreatedBy}} docker.io/library/ruby:latest
```
Possible output:

```dockerfile
CMD ["irb"]
RUN /bin/sh -c mkdir -p "$GEM_HOME" && chmod 1777 "$GEM_HOME" # buildkit
ENV PATH=/usr/local/bundle/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
ENV BUNDLE_SILENCE_ROOT_WARNING=1 BUNDLE_APP_CONFIG=/usr/local/bundle
ENV GEM_HOME=/usr/local/bundle
RUN /bin/sh -c set -eux;                savedAptMark="$(apt-mark showmanual)";  apt-get update;         apt-get install -y --no-install-recommends              dpkg-dev                libgdbm-dev            ruby    ;       rm -rf /var/lib/apt/lists/*;            rustArch=;      dpkgArch="$(dpkg --print-architecture)";        case "$dpkgArch" in             'amd64') rustArch='x86_64-unknown-linux-gnu'; rustupUrl='https://static.rust-lang.org/rustup/archive/1.26.0/x86_64-unknown-linux-gnu/rustup-init'; rustupSha256='0b2f6c8f85a3d02fde2efc0ced4657869d73fccfce59defb4e8d29233116e6db' ;;                 'arm64') rustArch='aarch64-unknown-linux-gnu'; rustupUrl='https://static.rust-lang.org/rustup/archive/1.26.0/aarch64-unknown-linux-gnu/rustup-init'; rustupSha256='673e336c81c65e6b16dcdede33f4cc9ed0f08bde1dbe7a935f113605292dc800' ;;        esac;           if [ -n "$rustArch" ]; then             mkdir -p /tmp/rust;                             wget -O /tmp/rust/rustup-init "$rustupUrl";            echo "$rustupSha256 */tmp/rust/rustup-init" | sha256sum --check --strict;               chmod +x /tmp/rust/rustup-init;                       export RUSTUP_HOME='/tmp/rust/rustup' CARGO_HOME='/tmp/rust/cargo';              export PATH="$CARGO_HOME/bin:$PATH";            /tmp/rust/rustup-init -y --no-modify-path --profile minimal --default-toolchain '1.74.1' --default-host "$rustArch";                           rustc --version;                cargo --version;        fi;             wget -O ruby.tar.xz "$RUBY_DOWNLOAD_URL";      echo "$RUBY_DOWNLOAD_SHA256 *ruby.tar.xz" | sha256sum --check --strict;                 mkdir -p /usr/src/ruby;         tar -xJf ruby.tar.xz -C /usr/src/ruby --strip-components=1;   rm ruby.tar.xz;          cd /usr/src/ruby;               {               echo '#define ENABLE_PATH_CHECK 0';             echo;           cat file.c;     } > file.c.new;         mv file.c.new file.c;          autoconf;       gnuArch="$(dpkg-architecture --query DEB_BUILD_GNU_TYPE)";      ./configure             --build="$gnuArch"              --disable-install-doc           --enable-shared                ${rustArch:+--enable-yjit}      ;       make -j "$(nproc)";     make install;           rm -rf /tmp/rust;       apt-mark auto '.*' > /dev/null;         apt-mark manual $savedAptMark > /dev/null;     find /usr/local -type f -executable -not \( -name '*tkinter*' \) -exec ldd '{}' ';'             | awk '/=>/ { so = $(NF-1); if (index(so, "/usr/local/") == 1) { next }; gsub("^/(usr/)?", "", so); printf "*%s\n", so }'              | sort -u               | xargs -r dpkg-query --search          | cut -d: -f1           | sort -u               | xargs -r apt-mark manual     ;       apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false;           cd /;   rm -r /usr/src/ruby;    if dpkg -l | grep -i ruby; then exit 1; fi;    [ "$(command -v ruby)" = '/usr/local/bin/ruby' ];       ruby --version;         gem --version;  bundle --version # buildkit
ENV RUBY_DOWNLOAD_SHA256=0686941a3ec395a15ae2a852487b2a88e5fb8a5518e188df00d8d1bb71a6349b
ENV RUBY_DOWNLOAD_URL=https://cache.ruby-lang.org/pub/ruby/3.3/ruby-3.3.1.tar.xz
ENV RUBY_VERSION=3.3.1
ENV LANG=C.UTF-8
RUN /bin/sh -c set -eux;        mkdir -p /usr/local/etc;        {               echo 'install: --no-document';          echo 'update: --no-document';   } >> /usr/local/etc/gemrc # buildkit
/bin/sh -c set -ex;     apt-get update;         apt-get install -y --no-install-recommends              autoconf                automake                bzip2           default-libmysqlclient-dev             dpkg-dev                file            g++             gcc             imagemagick             libbz2-dev              libc6-dev               libcurl4-openssl-dev          libdb-dev                libevent-dev            libffi-dev              libgdbm-dev             libglib2.0-dev          libgmp-dev              libjpeg-dev             libkrb5-dev           liblzma-dev              libmagickcore-dev               libmagickwand-dev               libmaxminddb-dev                libncurses5-dev                 libncursesw5-dev                libpng-dev             libpq-dev               libreadline-dev                 libsqlite3-dev          libssl-dev              libtool                 libwebp-dev             libxml2-dev           libxslt-dev              libyaml-dev             make            patch           unzip           xz-utils                zlib1g-dev      ;       rm -rf /var/lib/apt/lists/*
/bin/sh -c set -eux;    apt-get update;         apt-get install -y --no-install-recommends              git             mercurial               openssh-client          subversion            procps   ;       rm -rf /var/lib/apt/lists/*
/bin/sh -c set -eux;    apt-get update;         apt-get install -y --no-install-recommends              ca-certificates                 curl            gnupg           netbase               sq               wget    ;       rm -rf /var/lib/apt/lists/*
/bin/sh -c #(nop)  CMD ["bash"]
/bin/sh -c #(nop) ADD file:b9a9fc37b874060c713002ae1ac220f097edd7c6576116c22bb15aad8229b1b3 in / 

```

**Comment by LanikSJ** [comment]
The wonders of mobile notifications. 😀 

I'll add it to the docs.
